zkchannels-crypto/examples/range_proof.rs

```diff
@@ -0,0 +1,161 @@
+use {
+    bls12_381::{G2Projective, Scalar},
+    thiserror::Error,
+};
+
+use rand::thread_rng;
+use zkchannels_crypto::{
+    pedersen::PedersenParameters,
+    proofs::{
+        ChallengeBuilder, CommitmentProof, CommitmentProofBuilder, RangeConstraint,
+        RangeConstraintBuilder, RangeConstraintParameters, ValueOutsideRange,
+    },
+    BlindingFactor, Message, Rng,
+};
+
+fn main() {
+    // Generate parameters
+    let mut rng = thread_rng();
+    let pedersen_parameters = PedersenParameters::new(&mut rng);
+    let range_constraint_parameters = RangeConstraintParameters::new(&mut rng);
+
+    // Make a proof that your random number is in the range
+    let (blinding_factor, random_number_proof) = match RangeProof::new(
+        &mut rng,
+        &pedersen_parameters,
+        &range_constraint_parameters,
+        12345,
+    ) {
+        Ok(proof) => proof,
+        Err(e) => {
+            eprintln!("Impossible error while generating proof: {}", e);
+            return;
+        }
+    };
+
+    // Anyone can verify the proof without knowing your number
+    match random_number_proof.verify(&pedersen_parameters, &range_constraint_parameters) {
+        Ok(()) => println!("Yay! Proof verified!"),
+        Err(e) => eprintln!("Invalid zk proof! {}", e),
+    };
+
+    // You can't make a proof on a number that isn't in the range...
+    match RangeProof::new(
+        &mut rng,
+        &pedersen_parameters,
+        &range_constraint_parameters,
+        -100,
+    ) {
+        Ok(_proof) => eprintln!("Impossible error generating range proof on an invalid value!"),
```

Comment:
If these are truly impossible, you could mark them `unreachable!(...)` https://doc.rust-lang.org/std/macro.unreachable.html ?